src/primaite/agents/aegis/gllm.py

```diff
@@ -39,105 +37,92 @@ def __init__(
+    ) -> List[str]:
         # Format gllm prompt. We don't format context content as this is the graph embeddings we append to the embeddings of this prompt.
         if model == "openai":
-            prompt = GLLM_PROMPT.format(question=question, context=network_desc)
+            prompts = [GLLM_PROMPT.format(question=question, context=network_desc) for question in questions]
         elif model == "gllm":
-            prompt = GLLM_PROMPT.format(question=question, context="")
-        return prompt
-    
-    def get_cosine_embeddings_loss(self, openai_response, gllm_response):
-
-        openai_tokens = self.llm.tokenizer.batch_encode_plus(openai_response, return_tensors="pt", padding=True).input_ids
-        gllm_tokens = self.llm.tokenizer.batch_encode_plus(gllm_response, return_tensors="pt", padding=True).input_ids
-        
-        openai_tokens, gllm_tokens = self._pad_smallest_texts(openai_tokens=openai_tokens, gllm_tokens=gllm_tokens)
-
-        openai_response_embs = self.llm.get_embeddings(token_ids=openai_tokens, apply_chat_tokens=False, grad=True).view(-1, self.llm.llm_embedding_size)
-        gllm_response_embs = self.llm.get_embeddings(token_ids=gllm_tokens, apply_chat_tokens=False, grad=True).view(-1, self.llm.llm_embedding_size)
-        # try:
+            prompts = [GLLM_PROMPT.format(question=question, context="") for question in questions]
+
+        return prompts
+
+    def get_cosine_embeddings_loss(self, openai_hidden_states, gllm_hidden_states):
+
         loss = self.loss_fn(
-            gllm_response_embs, openai_response_embs, torch.ones(openai_response_embs.shape[0]).to("cuda:0")
+            openai_hidden_states.view(-1, self.llm.llm_embedding_size).to(gllm_hidden_states.device),
+            gllm_hidden_states.view(-1, self.llm.llm_embedding_size),
+            torch.ones(gllm_hidden_states.shape[0]).to(gllm_hidden_states.device),
         )
-        # except:
-        #     loss = -1
 
         return loss
 
-    def _pad_smallest_texts(self, openai_tokens, gllm_tokens):
-        if openai_tokens.shape[-1] > gllm_tokens.shape[-1]:
-            n_pad = openai_tokens.shape[-1] - gllm_tokens.shape[-1]
-            gllm_tokens = F.pad(gllm_tokens, (0, n_pad), "constant", self.llm.tokenizer.pad_token_id)
-        elif gllm_tokens.shape[-1] > openai_tokens.shape[-1]:
-            n_pad = gllm_tokens.shape[-1] - openai_tokens.shape[-1]
-            openai_tokens = F.pad(openai_tokens, (0, n_pad), "constant", self.llm.tokenizer.pad_token_id)
 
-        return openai_tokens, gllm_tokens
+def train_loop(model: GLLM, dataloader: DataLoader, network_desc: str, n_epochs: int = 10):
+    model.train(mode=True)
 
+    for epoch in range(n_epochs):
+        gllm_responses = []
 
-def train_loop(model, x, edge_index, questions: List[str], network_desc: str):
+        for batch in dataloader:
+            model.optimizer.zero_grad()
```

Comment:
`zero_grad()` needs to be moved out of the batch loop or `model.optimizer.step()` moved in.

Reply:
moved optimizer step in - good spot :D